### PR TITLE
Feature/wf 447

### PR DESCRIPTION
**Pull request by silvanbaumann**
## Pull request
Adds 'e-input'  component with several state options.
 
### Ticket
https://winterhalter-fenner.atlassian.net/browse/WF-447
 
### Browser testing
http://localhost:8080/styleguide/forms
http://localhost:6060/styleguide#einput
 
### Checklist
- [x] Tested all links in project relevant browsers
- [x] Tested all links in different screen sizes
- [x] Did run automated tests and linters
- [x] Did assign ticket
- [x] Double checked target branch
 
## Review/Test checklist
- [x] Did run automated tests and linters
- [x] Did review code
- [x] Tested all links in project relevant browsers
- [x] Tested all links in different screen sizes
- [x] Re-assign ticket to developer

### CODE REVIEW

**Comment by silvanbaumann** [comment]
Browser review

* Please add viewport meta to `index.html` according to the content-wrapper branch (`<meta name="viewport" content="width=device-width, initial-scale=1">`) for mobile testing.
[x] done - allowed to zoom? https://cl.ly/qwv1

* [ ] Input is 33px (Chrome, Edge)/32px (IE11 Win7/Win8) height. Should be 30px. Seems to be correct in IE11 on Win10
[x] done - added height-property to input field (30px)

* Please make :active shadow a bit larger according to design (also on sides).
[x] done

* IE11 shows a "delete" button on the right side of the field. This could become a problem with the state indicator later on.
[x] done

* Text position looks a bit low on IE11. Probably adjust vertical padding slightly but I think it's also OK to use it that way. (Edge is OK)
[ ] @patric-eberle is natural behavior. Shall we really add some IE hacks?

* On iOS there are still some browser styles applied (e.g. inner shadow). See 
http://wiki.derverstaerker.ch/css-basics/#-4
[x] done

* Color value for input text and placeholder is not applied in macOS/iOS Safari (in component style guide)
[x] done

* When moving mouse out of :hover, :focus demo in component style guide, the value is cleared (maybe because you don't assign a value?)
[ ] done

* Please add TODO notes in component styleguide for error, success and info state.
[x] done

* Safari shows some kind of input helper on error, success and info demo on the left side. This could become an issue with the status indicator later on.
[x] done

* Disabled style not according to design on Chrome/Android, Win10/Edge
[x] done

* On iOS the browser zooms when focusing an input field in component styleguide. Please check if still an issue after you added the viewport meta.
[x] done - added meta https://cl.ly/qwv1

* When a field has a state the focus should add the border color also to the gap (e.g. on error the field gets a red border all around on focus.
[x] done - https://cl.ly/qyJT

  * bildschirmfoto 2018-04-17 um 15 16 05

File: app/components/e-input.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+#### `default`
+```
+<template>
+  <div>
+  
+    <e-input v-model="demo" name="demo" placeholder="enter text..."/> 
+    
+    <!-- demo only -->
+    <div v-if="demo"><br/>
```
[the remark this answers is not in the document]
Comment (silvanbaumann) [reply app/components/e-input.md]: Removed 'break' and tried to add spacing-class (spacing-top-30). Didn't work for me.

File: app/components/e-input.vue [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    }
+    // render() {},
+  };
+</script>
+
+<style lang="scss">
+  .e-input {
+    @include half-border($color-grayscale--500);
+
+    border: 1px solid transparent;
```
[the remark this answers is not in the document]
Comment (silvanbaumann) [reply app/components/e-input.vue]: Was needed for testing. Removed now, thank you for that!

File: app/styleguide/routes/forms.vue [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  <div :class="b()">
+    <h1>Form example</h1>
+
+    <h2>e-input</h2>
+    <p>excluding label</p>
+    <form>
+      first name: <e-input v-model="name.firstname" name="firstname" placeholder="..." />
+      last name: <e-input v-model="name.lastname" name="lastname" placeholder="..." />
+      <p v-if="name.firstname && name.lastname">Your name is: <strong>{{ name.firstname }} {{ name.lastname }}</strong></p>
+      <pre>{{ name }}</pre>
```
[the remark this answers is not in the document]
Comment (silvanbaumann) [reply app/styleguide/routes/forms.vue]: Thought it's much nicer to hide the paragraph by default and show it only, when a text is entered. Just for demonstration purpose. ;-) Remove or leave?

File: app/components/e-input.vue [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    margin-bottom: $spacing--10;
+    position: relative;
+
+    // input
+    &__field {
+      border: 1px solid transparent;
+      border-radius: 2px;
```
[the remark this answers is not in the document]
Comment (silvanbaumann) [reply app/components/e-input.vue]: Glad, that we now have a variable.  ;-)

File: app/components/e-input.md [hunk 5]
```diff
@@ -8,7 +6,7 @@
     <e-input v-model="demo" name="demo" placeholder="enter text..."/> 
     
     <!-- demo only -->
-    <div v-if="demo"><br/>
+    <div v-if="demo" class="spacing--top-15">
```
Comment (silvanbaumann) [inline app/components/e-input.md]: I think this is not working properly. Have I done a mistake?

Comment (patric-eberle) [reply app/components/e-input.md]: Bug in the mixin. Fixed it.

**Comment by silvanbaumann** [comment]
As discussed 'zoom' is now disabled within `<meta ...` to prevent focus behavior in iOS.

File: app/components/e-input.vue [hunk 6]
```diff
@@ -0,0 +1,221 @@
+<!-- This component has no <template> because of dynamic root element -->
+
+<template>
+
+  <div :class="b(modifiers)">
+    <input :value="value"
+           :class="b('field')"
+           :disabled="disabled"
+           :placeholder="$attrs.placeholder"
```
[the remark this answers is not in the document]
Comment (silvanbaumann) [reply app/components/e-input.vue]: Added, we should discuss this maybe in the next frontend meeting. (default is now `off`)

**Comment by silvanbaumann** [comment]
Didn't see the expanded comments. Should be done now. thx for the hint.

File: app/setup/scss/mixins/_half-border.scss [hunk 7]
```diff
@@ -0,0 +1,42 @@
+////
+// Adds half-border to the element
+//
+// @param  {String}   Color of the border
+//
+////
+
+@mixin half-border($border-color: $color-grayscale--500) {
+  @if $border-color != null {
```
Comment (patric-eberle) [inline app/setup/scss/mixins/_half-border.scss]: - [ ] Please remove condition. Not needed anymore, since you added a fallback.

**Comment by patric-eberle** [comment]
@silvanbaumann there is also a linting error. Please run `npm run test` and fix it.

**Comment by silvanbaumann** [comment]
@patric-eberle Line removed.

File: app/setup/scss/mixins/_half-border.scss (hunk 7 above)
Comment (silvanbaumann) [reply app/setup/scss/mixins/_half-border.scss]: done

**Pull request merged**